### PR TITLE
wiki: add get-started guide for Go fuzzing

### PR DESCRIPTION
**Pull request by ukane-philemon**
This is a get-started guide that provides necessary information on how to start writing Go fuzz tests. I've tried to put all the necessary information in one place and added references to the Go docs. Part of #1748

### CODE REVIEW

**Comment by chappjc** [comment]
Great start!  Thanks for the docs and the good first use case.

File: docs/wiki/Getting Started With Fuzzing.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# Getting Started With Fuzzing
+
+Go started supporting fuzzing in its standard toolchain beginning in Go 1.18. Fuzzing is a type of automated testing which continuously manipulates inputs to a program to find issues such as panics or bugs. These semi-random data mutations can discover new code coverage that existing unit tests may miss, and uncover edge case bugs which would otherwise go unnoticed. Since fuzzing can reach these edge cases, fuzz testing is particularly valuable for finding security exploits and vulnerabilities.
```
Comment (chappjc) [inline docs/wiki/Getting Started With Fuzzing.md:3]: Let's link to https://go.dev/doc/fuzz/ in this paragraph.

File: dex/encrypt/encrypt_test.go [hunk 2]
```diff
@@ -72,6 +72,35 @@ func TestDecrypt(t *testing.T) {
...
+			t.Fatalf("Encrypt error: %v", err)
+		}
+		reThing, err := crypter.Decrypt(encThing)
+		if err != nil {
+			t.Fatalf("Decrypt error: %v", err)
+		}
+		if !bytes.Equal(thing, reThing) {
+			t.Fatalf("%x != %x", thing, reThing)
+		}
+	})
```
Comment (chappjc) [inline dex/encrypt/encrypt_test.go:101]: Got a panic, negative slice length it seems:

```
=== FUZZ  FuzzDecrypt
fuzz: elapsed: 0s, gathering baseline coverage: 0/4 completed
fuzz: elapsed: 0s, gathering baseline coverage: 4/4 completed, now fuzzing with 32 workers
fuzz: minimizing 45-byte failing input file
fuzz: elapsed: 2s, minimizing
--- FAIL: FuzzDecrypt (1.99s)
    --- FAIL: FuzzDecrypt (0.05s)
        testing.go:1356: panic: runtime error: makeslice: len out of range
            goroutine 513 [running]:
            runtime/debug.Stack()
            	/home/jon/go119/src/runtime/debug/stack.go:24 +0xdb
            testing.tRunner.func1()
            	/home/jon/go119/src/testing/testing.go:1356 +0x1f2
            panic({0x62a620, 0x695060})
            	/home/jon/go119/src/runtime/panic.go:884 +0x212
            decred.org/dcrdex/dex/encode.RandomBytes(0xc000114220?)
            	/home/jon/github/decred/dcrdex/dex/encode/encode.go:65 +0x65
            decred.org/dcrdex/dex/encrypt.FuzzDecrypt.func1(0xc000382340, 0xffffffffffffffc3, {0xc000114220, 0x1, 0x10})
            	/home/jon/github/decred/dcrdex/dex/encrypt/encrypt_test.go:92 +0x167
            reflect.Value.call({0x626920?, 0x6629b8?, 0x13?}, {0x653599, 0x4}, {0xc00011a1e0, 0x3, 0x4?})
            	/home/jon/go119/src/reflect/value.go:584 +0x8c5
            reflect.Value.Call({0x626920?, 0x6629b8?, 0x51b?}, {0xc00011a1e0?, 0x761520?, 0x7838c0?})
            	/home/jon/go119/src/reflect/value.go:368 +0xbc
            testing.(*F).Fuzz.func1.1(0x1fbefaf4a4f5b574?)
            	/home/jon/go119/src/testing/fuzz.go:337 +0x231
            testing.tRunner(0xc000382340, 0xc0001f0090)
            	/home/jon/go119/src/testing/testing.go:1446 +0x10b
            created by testing.(*F).Fuzz.func1
            	/home/jon/go119/src/testing/fuzz.go:324 +0x5b9
            
    
    Failing input written to testdata/fuzz/FuzzDecrypt/465a95d340d5bc76fcb0352a6608e97a0ca215efcabd41010b86359b8f7d3cd1
    To re-run:
    go test -run=FuzzDecrypt/465a95d340d5bc76fcb0352a6608e97a0ca215efcabd41010b86359b8f7d3cd1
FAIL
exit status 1
FAIL	decred.org/dcrdex/dex/encrypt	2.272s
```

```
$  cat testdata/fuzz/FuzzDecrypt/465a95d340d5bc76fcb0352a6608e97a0ca215efcabd41010b86359b8f7d3cd1
go test fuzz v1
int(-61)
[]byte("0")
```

File: docs/wiki/Getting Started With Fuzzing.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+ - With fuzzing `go test -fuzz=FuzzTestName`.
+    Each seed corpus entry will be tested against the fuzz target. Also, If there are seed corpus already generated for the fuzz target in `$GOCACHE` they will be run against the fuzz target, and then the fuzzing engine will continue to generate and test seed corpus until `-fuzztime` (See [Fuzz Flags](#Other-Highlighted-Fuzz-Flags)).
+    
+    To enable fuzzing, run go test with the `-fuzz` flag, providing a test name (e.g `FuzzTestName`) or regex matching a single fuzz test. By default, all other tests in that package will run before fuzzing begins. This is to ensure that fuzzing won’t report any issues that would already be caught by an existing test.
+
+    **Note**: Fuzzing cannot be run for multiple packages at the same time using the `-fuzz` flag. You must specify at most one Fuzz test (e.g `-fuzz=FuzzXXX`) (See [suppport for multiple Fuzz tests](https://go.googlesource.com/proposal/+/master/design/draft-fuzzing.md#fuzzing-engine-supports-multiple-fuzz-tests-at-once)).
```
Comment (chappjc) [inline docs/wiki/Getting Started With Fuzzing.md:73]: From "To enable..." down to here is indented as part of the previous bullet.  Once you get markdownlint running and remove stray spaces it should start to look right.

**Review by chappjc — CHANGES_REQUESTED** [review]

File: dex/encrypt/encrypt_test.go [hunk 4]
```diff
@@ -72,6 +72,35 @@ func TestDecrypt(t *testing.T) {
...
+func FuzzDecrypt(f *testing.F) {
+	seeds := [][]byte{
+		[]byte("4kliaOCha2"),
+		[]byte("123456"),
+		[]byte("23Fgfge34"),
+		[]byte("asdfgth#4"),
+	}
+
+	for _, seed := range seeds {
+		f.Add(rand.Intn(100), seed)
```
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply dex/encrypt/encrypt_test.go:84]: > I'm curious how the value of the integer and the length of the seed byte sliced affects the values the fuzzer feeds to the ff func.

The value of the integer and length of the seed does affect seeds generated by the fuzzer. It needs sensible seeds to replicate the same. But even if we at times get undesired seeds, we can always use a filter to skip seeds that are sure to fail(malformed/bad seeds).

File: docs/wiki/Getting Started With Fuzzing.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
+# Getting Started With Fuzzing
+
+Go started supporting fuzzing in its standard toolchain beginning in Go 1.18. Fuzzing is a type of automated testing which continuously manipulates inputs to a program to find issues such as panics or bugs. These semi-random data mutations can discover new code coverage that existing unit tests may miss, and uncover edge case bugs which would otherwise go unnoticed. Since fuzzing can reach these edge cases, fuzz testing is particularly valuable for finding security exploits and vulnerabilities.
+
+## Writing a Fuzz Test
```
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply docs/wiki/Getting Started With Fuzzing.md:5]: > In fact, let's create a ".markdownlint.json" at the repo root:

Are there more rules we'd like to add?

File: docs/wiki/Getting Started With Fuzzing.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+// `f.Fuzz` accepts a function and runs fuzzing with randomly generated values 
+// based on the seed corpus provided. All seed corpus entries must have types
+// which are identical to the fuzzing arguments, in the same order. This is true
+// for calls to `(*testing.F).Add`.
+	f.Fuzz(func(t *testing.T, n int, a []byte) {
+		// Inputs can be validated and skipped if they are not suitable.
+		if n > 100000000 {
+			t.Skip()
+		}
```
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply docs/wiki/Getting Started With Fuzzing.md:38]: > Does the fuzzer have bounds

No, it doesn't. 

> Is it related to the seeds?

`a`'s length is random. The fuzzing engine only needs to know the parameter type and will generate new or mutate provided seeds.

**Comment by chappjc** [comment]
Hang on.  😊. Narrow the scope back up.  We can get to a global lint policy later.  Just fuzzing now please

**Comment by ukane-philemon** [comment]
Fuzzed for 8min+,  31 `new interesting seeds generated.
Screenshot 2022-08-18 at 3 57 47 PM

**Comment by ukane-philemon** [comment]
> And then fix anything it's still complaining about.
> Hang on. 😊. Narrow the scope back up. We can get to a global lint policy later. Just fuzzing now please

😅 `markdownlint` kept complaining.

**Comment by chappjc** [comment]
> > And then fix anything it's still complaining about.
> 
> > Hang on. 😊. Narrow the scope back up. We can get to a global lint policy later. Just fuzzing now please
> 
> 😅 `markdownlint` kept complaining.
> 
> EDIT: Should the global lint fix come in a separate PR?

Totally.  We can really dig into that, CI and all, in a follow up PR.

I get it though, it's like whackamole

File: package.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "dcrdex",
+  "version": "1.1.0",
+  "description": "[![Build Status](https://github.com/decred/dcrdex/workflows/Build%20and%20Test/badge.svg)](https://github.com/decred/dcrdex/actions) [![ISC License](https://img.shields.io/badge/license-Blue_Oak-007788.svg)](https://blueoakcouncil.org/license/1.0.0) [![GoDoc](https://img.shields.io/badge/go.dev-reference-blue.svg?logo=go&logoColor=lightblue)](https://pkg.go.dev/decred.org/dcrdex)",
+  "private": true,
+  "scripts": {
+    "lint": "markdownlint docs/wiki/*.md && markdownlint docs/release-notes/*.md"
```
Comment (chappjc) [inline package.json:7]: Eh, this doesn't seem right.  I don't think we should make an npm package just to run a linter, especially not at the root of the repo and named "dcrdex".

I guess I kinda implied we'd need some kind of CI solution, but I just run markdownlint one of two ways: (a) VSCode extension, or (b) system-wide install of markdownlint-cli.

Let's figure out enforcement of markdownlint policy later, and just use our tentative policy we've discussed for the fuzzing markdown.

**Comment by chappjc** [comment]
Looks, good.  One more thing - let's add `**/testdata/fuzz` to .gitignore

File: docs/wiki/Getting Started With Fuzzing.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+- Keeping auto-generated seed corpus in $GOCACHE increases the efficiency of the fuzzing engine. The more seed corpus the better.
+- Provide sufficient seed corpus. This could be in the test file, another file or a directory. This will be a guide to the fuzzing engine in generating sensible random inputs.
+To convert provided seed corpus files to Go fuzzing corpus format use [file2fuzz](https://pkg.go.dev/golang.org/x/tools/cmd/file2fuzz):
+
+  ```sh
+
+	$ go install golang.org/x/tools/cmd/file2fuzz@latest
+	$ file2fuzz
+
+  ```
```
Comment (JoeGruffins) [inline docs/wiki/Getting Started With Fuzzing.md:146]: What do I do with file2fuzz?

File: docs/wiki/Getting Started With Fuzzing.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+**Note**: Fuzzing cannot be run for multiple packages at the same time using the `-fuzz` flag. You must specify at most one Fuzz test (e.g `-fuzz=FuzzXXX`) [See: [suppport for multiple Fuzz tests](https://go.googlesource.com/proposal/+/master/design/draft-fuzzing.md#fuzzing-engine-supports-multiple-fuzz-tests-at-once)].
+
+## Command Line Output
+
+- **elapsed**: the amount of time that has elapsed since the process began
+- **execs**: the total number of inputs that have been run against the fuzz target (with an average execs/sec since the last log line)
+- **new interesting**: the total number of “interesting” inputs that have been added to the generated corpus during this fuzzing execution (with the total size of the entire corpus)
+
+```text
+philemon@Philemons-MBP encrypt % go test -v ./... --fuzz=FuzzDecrypt --fuzztime=10x
```
Comment (JoeGruffins) [inline docs/wiki/Getting Started With Fuzzing.md:97]: Maybe better to take out the name here, just $ maybe?
```
$ go test -v --fuzz=FuzzDecrypt --fuzztime=10x
```

**Review by JoeGruffins — APPROVED** [review]

File: docs/wiki/Getting Started With Fuzzing.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+- **-parallel**: the number of fuzzing processes running at once, default $GOMAXPROCS. Currently, setting -cpu during fuzzing has no effect.
+- **-keepfuzzing**: Keep running the fuzz test if a crasher is found. (default false)
+
+## Best Practices
+
+- Fuzz targets `should be fast and deterministic` so the fuzzing engine can work efficiently, and new failures and code coverage can be easily reproduced.
+ Fuzzing is most effective with the most primitive methods and functions. As such, this will take some careful targeting and potentially some very light refactoring here and there to get fuzzable functions.
+- Since the fuzz target is invoked in parallel across multiple workers and in nondeterministic order, the state of a fuzz target should not persist past the end of each call, and the behaviour of a fuzz target `should not depend on a global state`.
+- Keeping auto-generated seed corpus in $GOCACHE increases the efficiency of the fuzzing engine. The more seed corpus the better.
```
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply docs/wiki/Getting Started With Fuzzing.md:137]: >  do not do f.Clean(f func())

No, best practices ain't hard rules. Besides, `f.Clean(f func())` is not tied to cleaning up seed corpus.

File: docs/wiki/Getting Started With Fuzzing.md (hunk 8 above)
Comment (ukane-philemon) [reply docs/wiki/Getting Started With Fuzzing.md:146]:
```text
To convert provided seed corpus files to Go fuzzing corpus format use [file2fuzz]
```

Comment (chappjc) [reply docs/wiki/Getting Started With Fuzzing.md]: TBH this still isn't clear to me.  I can see that file2fuzz takes some binary file as input and outputs to the "Go fuzzing corpus format".  But what kind of input fie?  Then where do you put the output and how do you use it?

Comment (chappjc) [reply docs/wiki/Getting Started With Fuzzing.md]: `file2fuzz` sounds like an alternative to using a random string of your choice to go in the `[]byte`.  Basically it puts the contents of the file into a corpus like:

```
go test fuzz v1
[]byte("<contents of file>")
```

And then you... ? put that somewhere for the go fuzz test to find and use?

Comment (ukane-philemon) [reply docs/wiki/Getting Started With Fuzzing.md]: > And then you... ? put that somewhere for the go fuzz test to find and use?

Yes, `file2fuzz` has an `-o` flag that provides a path to write output(seed corpus in Go fuzzing corpus format) files to.
